### ISSUE TITLE
Hang when doing go-to-def

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#40455]
Uploading dump now.

**Comment by CyrusNajmabadi** [comment]
Tagging @jinujoseph @mavasani

**Comment by mavasani** [comment]
Tagging @allisonchou as well. We both have also been seeing this often in latest preview builds. @CyrusNajmabadi let me know if you have a dump, otherwise, I will collect one when I hit this issue. Also are you able to repro this with consistent steps?

**Comment by CyrusNajmabadi** [comment]
No.  I have no repro on any of this stuff.  I'm just using 16.5 p1 (and it's super unpleasant).

**Comment by CyrusNajmabadi** [comment]
@mavasani Dump is here: https://drive.google.com/file/d/1rUmMcGWMX47z6PSyHnTeVeHD6pSt6qnD/view?usp=sharing

**Comment by mavasani** [comment]
Seems like the block is coming from `Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open`

@jasonmalinowski @davkean Should this bug be ported to https://github.com/dotnet/project-system or internal CPS repo?

**Main thread:**

``` 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.NoMessagePumpSyncContext.Wait(System.IntPtr[] waitHandles, bool waitAll, int millisecondsTimeout)	Unknown
 	mscorlib.dll!System.Threading.SynchronizationContext.InvokeWaitMethodHelper(System.Threading.SynchronizationContext syncContext, System.IntPtr[] waitHandles, bool waitAll, int millisecondsTimeout) Line 349	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext) Line 203	C#
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) Line 213	C#
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 688	C#
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3327	C#
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3259	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3167	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Wait(System.TimeSpan timeout) Line 3083	C#
 	[Waiting on Async Operation, Open Tasks View in Parallel Stacks Window to view Async Call Stacks]	
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.DelegatingJoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.Join(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<bool>.Join(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__3()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__1(object sender, Microsoft.VisualStudio.Shell.BlockingTaskEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.RaiseOnMarkedAsBlockingEvent(Microsoft.VisualStudio.Shell.Interop.IVsTask blockedTask)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog<System.__Canon>(System.Func<System.__Canon> function)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog(System.Action action)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalGetResult(bool ignoreUIThreadCheck)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetResult()	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, bool openWith, uint editorFlags, ref System.Guid editorType, string physicalView, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, bool openWith, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.OpenItem.AnonymousMethod__0()	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.dll!Microsoft.VisualStudio.ProjectSystem.VS.HResult.Invoke(System.Func<Microsoft.VisualStudio.ProjectSystem.VS.HResult> action, System.IServiceProvider vsShellServiceProvider, Microsoft.VisualStudio.ProjectSystem.IProjectFaultHandlerService projectFaultHandlerService, Microsoft.VisualStudio.ProjectSystem.UnconfiguredProject project)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.OpenItem(uint itemId, ref System.Guid logicalViewParam, System.IntPtr punkDocDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame frameParam)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNodeWrapper.OpenItem(uint itemid, ref System.Guid rguidLogicalView, System.IntPtr punkDocDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ppWindowFrame)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.TryGetFrame(Microsoft.CodeAnalysis.Document document, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame frame) Line 1097	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.OpenDocumentCore(Microsoft.CodeAnalysis.DocumentId documentId, bool activate) Line 1049	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.VisualStudioWorkspaceImpl.OpenDocument(Microsoft.CodeAnalysis.DocumentId documentId, bool activate) Line 1007	C#
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.VisualStudioDocumentNavigationService.OpenDocument(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.DocumentId documentId) Line 287	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.VisualStudioDocumentNavigationService.TryNavigateToSpan(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.DocumentId documentId, Microsoft.CodeAnalysis.Text.TextSpan textSpan, Microsoft.CodeAnalysis.Options.OptionSet options) Line 134	C#
 	Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.DocumentSpanExtensions.TryNavigateTo(Microsoft.CodeAnalysis.DocumentSpan documentSpan, bool isPreview) Line 21	C#
```

**Comment by allisonchou** [comment]
I'm not sure if a screenshot is helpful, but this is what has been happening constantly on my machine using 16.5 P1. The file eventually loads after a while, but it takes at least a few minutes.
![gotohang](https://user-images.githubusercontent.com/16968319/71044960-17adf780-20e8-11ea-842e-c09ef25619c2.png)

**Comment by davkean** [comment]
You are missing the native frames:

```
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task) Line 359	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task) Line 336	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.DelegatingJoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task) Line 48	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread() Line 881	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.Join(System.Threading.CancellationToken cancellationToken) Line 554	C#
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<bool>.Join(System.Threading.CancellationToken cancellationToken) Line 68	C#
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__3()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__1(object sender, Microsoft.VisualStudio.Shell.BlockingTaskEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.RaiseOnMarkedAsBlockingEvent(Microsoft.VisualStudio.Shell.Interop.IVsTask blockedTask)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog<System.__Canon>(System.Func<System.__Canon> function)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog(System.Action action)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalGetResult(bool ignoreUIThreadCheck)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetResult()	Unknown
 	[Native to Managed Transition]	
 	msenv.dll!CVsUIShellOpenDocument::DoSecurityCheck(const wchar_t * pszMkDocument) Line 9409	C++
 	msenv.dll!CVsUIShellOpenDocument::OpenStandardEditor(unsigned long grfOpenStandard, const wchar_t * pszMkDocument, const _GUID & rguidLogicalView, const wchar_t * pszOwnerCaption, IVsUIHierarchy * pHier, unsigned long itemid, IUnknown * punkDocDataExisting, IServiceProvider * pSP, IVsWindowFrame * * ppDocFrame) Line 8512	C++
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, bool openWith, uint editorFlags, ref System.Guid editorType, string physicalView, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction) Line 345	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, bool openWith, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction) Line 156	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode projectItem, ref System.Guid logicalView, System.IntPtr docDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame windowFrame, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction windowFrameAction) Line 59	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.OpenItem.AnonymousMethod__0() Line 1987	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.dll!Microsoft.VisualStudio.ProjectSystem.VS.HResult.Invoke(System.Func<Microsoft.VisualStudio.ProjectSystem.VS.HResult> action, System.IServiceProvider vsShellServiceProvider, Microsoft.VisualStudio.ProjectSystem.IProjectFaultHandlerService projectFaultHandlerService, Microsoft.VisualStudio.ProjectSystem.UnconfiguredProject project) Line 324	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.OpenItem(uint itemId, ref System.Guid logicalViewParam, System.IntPtr punkDocDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame frameParam) Line 1974	C#
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNodeWrapper.OpenItem(uint itemid, ref System.Guid rguidLogicalView, System.IntPtr punkDocDataExisting, out Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ppWindowFrame) Line 143	C#

```

This is not the project system. This is a 64-bit dump so can't use WinDBG to inspect async tasks, but I did eventually successfully look (if I left it for 15 minutes) at the async tasks using the new Tasks view in Parallel Stacks. This should be opened against the Shell team.

**Comment by mavasani** [comment]
Thanks @davkean. I have filed https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1042460 for the shell team. Closing this bug as external. I will provide an update once there is progress on the shell bug.

**Comment by olegtk** [comment]
something is fishy with IVsPathTrustVerifier, I've seen TypeScript reporting similar delay recently. Thanks for reporting and uploading a dump, @CyrusNajmabadi!

**Comment by olegtk** [comment]
Nope, wrong, it's likely the same issue as in 16.4, just fixed in 16.4.2 that shipped today: https://github.com/microsoft/vs-streamjsonrpc/issues/396. Good news it's already fixed in 16.5 Preview2 (but it ships only in January).

**Comment by CyrusNajmabadi** [comment]
> (but it ships only in January).

:'(
